src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -244,16 +218,115 @@ Status RedisStoreClient::AsyncMultiGet(
   RAY_CHECK(callback);
   if (keys.empty()) {
     callback({});
+    return Status::OK();
   }
   std::vector<std::string> true_keys;
   for (auto &key : keys) {
     true_keys.push_back(GenRedisKey(external_storage_namespace_, table_name, key));
   }
-  RAY_CHECK_OK(MGetValues(
-      redis_client_, external_storage_namespace_, table_name, true_keys, callback));
+  MGetValues(table_name, true_keys, callback);
   return Status::OK();
 }
 
+void RedisStoreClient::ProcessNext(std::string key) {
+  std::function<void()> op;
+  {
+    absl::MutexLock lock(&mu_);
+    auto op_iter = redis_ops_.find(key);
+    RAY_CHECK(op_iter != redis_ops_.end()) << " redis ops stats failed for key: " << key;
+    auto &queue = op_iter->second;
+    // All write operations have been consumed, remove this entry
+    if (queue.empty()) {
+      redis_ops_.erase(op_iter);
+    } else {
+      op = std::move(queue.front());
+    }
+  }
+  if (op) {
+    op();
+  }
+}
+
+void RedisStoreClient::SendRedisCmd(std::vector<std::string> keys,
+                                    std::vector<std::string> args,
+                                    RedisCallback redis_callback) {
+  RAY_CHECK(!keys.empty());
+
+  auto send_redis = [this,
+                     keys,
+                     args = std::move(args),
+                     redis_callback = std::move(redis_callback)]() mutable {
+    {
+      // Pop the operations from the queue because it's activately
+      // executing it.
+      absl::MutexLock lock(&mu_);
+      for (size_t i = 0; i < keys.size(); ++i) {
+        auto op_iter = redis_ops_.find(keys[i]);
+        RAY_CHECK(op_iter != redis_ops_.end());
+        auto &queue = op_iter->second;
+        RAY_CHECK(!queue.empty()) << " check redis ops failed for key: " << keys[i];
+        queue.pop();
+      }
+    }
+    // Send the actual request
+    auto cxt = redis_client_->GetShardContext("");
+    RAY_CHECK_OK(cxt->RunArgvAsync(
+        std::move(args),
+        [this, keys = std::move(keys), redis_callback = std::move(redis_callback)](
+            auto reply) {
+          // Fire the next request
+          for (const auto &key : keys) {
+            ProcessNext(key);
+          }
+          if (redis_callback) {
+            redis_callback(reply);
+          }
+        }));
+  };
+
+  bool can_fire = true;
+  // busiest key is defined to be the key has the most pending operations.
+  size_t busiest_key_idx = 0;
+  size_t busiest_key_queuing_size = 0;
+  {
+    absl::MutexLock lock(&mu_);
+    for (size_t i = 0; i < keys.size(); ++i) {
```

Comment:
would this recorder the request [ WRITE(A), Write(A, B), Write(B)]  into [Write(A), Write(B), Write(A,B) ]?

Reply:
No it wont I think.

The busiest_key_idx will be the place where the write will be pushed

Suppose we have 

A: []
B: []

Then, we have W(A)

A: [W(A)]
B: []

Then we have W(A, B), we'll have (suppose W(A) hasn't finished)

A: [W(A), W(A, B)]
B: [noop]

Then we have W(B), we'll have

A: [W(A), W(A, B)]
B: [noop, W(B)]  ===> since W(B) notice there is pending writes, W(B) won't be fired.


Then suppose W(A) finished, we'll have

A: [W(A, B)]
B: [noop, W(B)] ===> noop is still there because we never fire any ops for B

Then W(A, B) finished, we'll have

A:[]
B:[W(B)] ==> noop got evicted because W(A, B) is executed. And it'll start to execute W(B)

Then, we'll have

A:[]
B:[W(B)]

Reply:
makes sense